src/XtensaOptimize.cpp

```diff
@@ -0,0 +1,2331 @@
+        Expr call = Call::make(wild_u16x.type(), "halide_xtensa_concat_from_native",
+                               {std::move(v0), std::move(v1)}, Call::PureExtern);
+        return call;
+    }
+
+    static Expr halide_xtensa_concat_from_native_i24(Expr v0, Expr v1) {
+        Expr call = Call::make(wild_i24x.type(), "halide_xtensa_concat_from_native",
+                               {std::move(v0), std::move(v1)}, Call::PureExtern);
+        return call;
+    }
+
+    static Expr halide_xtensa_concat_from_native_i32(Expr v0, Expr v1) {
+        Expr call = Call::make(wild_i32x.type(), "halide_xtensa_concat_from_native",
+                               {std::move(v0), std::move(v1)}, Call::PureExtern);
+        return call;
+    }
+
+    static Expr halide_xtensa_concat_from_native_i32(Expr v0, Expr v1, Expr v2, Expr v3) {
+        Expr call = Call::make(wild_i32x.type(), "halide_xtensa_concat_from_native",
+                               {std::move(v0), std::move(v1), std::move(v2), std::move(v3)},
+                               Call::PureExtern);
+        return call;
+    }
+
+    static Expr halide_xtensa_concat_from_native_u32(Expr v0, Expr v1) {
+        Expr call = Call::make(wild_u32x.type(), "halide_xtensa_concat_from_native",
+                               {std::move(v0), std::move(v1)}, Call::PureExtern);
+        return call;
+    }
+
+    static Expr halide_xtensa_concat_from_native_u1(Expr v0, Expr v1, Expr v2, Expr v3) {
+        Expr call = Call::make(wild_u1x.type(), "halide_xtensa_concat_from_native",
+                               {std::move(v0), std::move(v1), std::move(v2), std::move(v3)}, Call::PureExtern);
+        return call;
+    }
+
+    static Expr halide_xtensa_concat_from_native_i48(Expr v0, Expr v1) {
+        Expr call = Call::make(wild_i48x.type(), "halide_xtensa_concat_from_native",
+                               {std::move(v0), std::move(v1)}, Call::PureExtern);
+        return call;
+    }
+
+    Expr visit(const Add *op) override {
+        if (op->type.is_vector()) {
+            static const std::vector<Pattern> adds = {
+                // Predicated addition
+                // NOTE(vksnk): patterns below are for predicated instructions and look like they may
+                // be more efficient, but they are not according to simulator. We will need to check with
+                // Cadence about this.
+                // {"halide_xtensa_pred_add_i8", wild_i8x + select(wild_u1x, wild_i8x, wild_i8x)},
+                // {"halide_xtensa_pred_add_i16", wild_i16x + select(wild_u1x, wild_i16x, wild_i16x)},
+                // {"halide_xtensa_pred_add_i32", wild_i32x + select(wild_u1x, wild_i32x, wild_i32x)},
+
+                {"halide_xtensa_qqqq", slice(wild_i24x256, 0, 1, 128) + slice(wild_i24x256, 128, 1, 128), Pattern::SameOp01},
+                {"halide_xtensa_yyyy", (call("halide_xtensa_xxxx", wild_i24x64, {wild_i24x64, wild_i24x128}) + slice(wild_i24x128, 64, 1, 64)), Pattern::SameOp12},
```

Comment:
Why are some methods implemented via the static call methods defined above, and some implemented via `call()`? imo the former is more readable, and the latter should be replaced by the former if possible.